src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,104 @@
-""" This is a dummy example """
+import requests
+import json
+import pandas as pd
+import matplotlib.pyplot as plt
+import numpy as np
+from datetime import timedelta
+from ratelimit import limits, sleep_and_retry
+
+
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+
+VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"
+
+
+def get_data_meteo_api(lat: float, long: float, variable: str):
+    # For the API call I only parametrized the coordinates as
+    # they are the only parameters that will vary in this case
+    API_URL_0 = f"https://climate-api.open-meteo.com/v1/climate?latitude={lat}&longitude={long}&start_date=1950-01-01&end_date=2050-12-31&models=CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S&daily="
+    API_URL = API_URL_0 + variable
+    return pd.DataFrame(json.loads(api_request_manager(API_URL)))
+
+
+# This function controls that the API call limit is not exceeded
+# and also raises an error if the status is not 200
+@sleep_and_retry
+@limits(calls=10000, period=timedelta(days=1).total_seconds())
+def api_request_manager(URL):
+    response = requests.get(URL)
+    response.raise_for_status()
+    return response.content.decode("utf-8")
+
+
+def calculate_mean_std(data: pd.DataFrame):
+    # Daily values for different models are put into a matrix to calculate mean and std
+    matrix = [np.array(data.iloc[0], dtype=np.float64)]
```

Comment:
I am quite confused about what are you doing in this function. First of all, please add a little explanation on how the df `data` that you have as a parameter looks like and what are the mean / std you want to compute. Once I know this, I can try to help you with the logic, but as a general rule is a bad idea to iterate over the rows of a data frame as you are doing

Reply:
The dataframe I got from the API looks like this 
![image](https://github.com/geeeb95/zrive-ds/assets/57015940/a2db6379-eaf5-4c98-8fe1-be59b358ad4e)
This means that there are only like 8 rows (one for the "time" and one for every climate model) and one column, so every single element in the dataframe is a series ~38k values. Ideally, I would prefer to have the current rows as columns and then every row would be an entry of the temperatures corresponding to a single date. I've tried many things (including pivoting the dataframe) but I couldn't get it to convert into a usable structure.
That is why I did this workaround: I turn the whole thing into a numpy array and this allows me to build a matrix with the desired structure where I can calculate mean and std. It's not ideal but it's the only way I found and since I only iterate through <10 dataframe rows maybe it's not that bad...
Also, I saw in Guille's solution that he didn't have this problem with the dataframe structure, but couldn't find out why.